sandstone/client/sandstone/components/filetreedirective/tests/filetree.directive.test.js

```diff
@@ -1,208 +1,421 @@
+            expect(node.children.length).toBe(3);
+        });
+
+        it('removeNodeFromFiletree: node exists', function(){
+            // Get contents of first node
+            isolateScope.getDirContents(isolateScope.treeData.filetreeContents[0], true);
+            // Remove the second node
+            var node = isolateScope.treeData.filetreeContents[0].children[1];
+            isolateScope.removeNodeFromFiletree(node);
+            expect(isolateScope.treeData.filetreeContents[0].children.length).toBe(2);
+        });
+
+        it('removeNodeFromFiletree: node does not exist', function(){
+            // expand the first node
+            isolateScope.getDirContents(isolateScope.treeData.filetreeContents[0], true);
+            // Some node that does not exist
+            var node = {
+                filepath: '/home/user/doesnotexist',
+                filename: 'doesnotexist',
+                type: 'file'
+            };
+            isolateScope.removeNodeFromFiletree(node);
+            // expect filetree to remain unchanged
+            expect(isolateScope.treeData.filetreeContents[0].children.length).toBe(3);
+        });
+
+        it('isExpanded:true', function(){
+            isolateScope.treeData.expandedNodes = [dirs[0]];
+            expect(isolateScope.isExpanded(dirs[0].filepath)).toBeTruthy();
+        });
+
+        it('isExpanded:false', function(){
+            expect(isolateScope.isExpanded(dirs[0].filepath)).not.toBeTruthy();
+        });
+
+        it('describeSelection', function(){
+            var node = isolateScope.treeData.filetreeContents[0];
+            isolateScope.describeSelection(node, true);
+            expect(isolateScope.treeData.selectedNodes.length).toBe(1);
+            isolateScope.describeSelection(node, false);
+            expect(isolateScope.treeData.selectedNodes.length).toBe(0);
+        });
+
+        it('$rootScope.$on: filtree:created_file', function() {
+            spyOn(isolateScope, 'updateFiletree');
+            rootScope.$emit('filetree:created_file', {
+                filepath: '/tmp/some-file.txt'
+            });
+            expect(isolateScope.updateFiletree).toHaveBeenCalled();
+        });
+        it('$rootScope.$on: filtree:deleted_file', function() {
+            spyOn(isolateScope, 'updateFiletree');
+            rootScope.$emit('filetree:deleted_file', {
+                filepath: '/tmp/some-file.txt'
+            });
+            expect(isolateScope.updateFiletree).toHaveBeenCalled();
+        });
+        it('$rootScope.$on: filetree:moved_file', function() {
+            spyOn(isolateScope, 'updateFiletree');
+            rootScope.$emit('filetree:moved_file', {
+                src_filepath: '/home/user/somefile',
+                dest_filepath: '/home/user/some_new_file'
+            });
+            expect(isolateScope.updateFiletree).toHaveBeenCalled();
+        });
+      });
+      describe('leafLevel: dir', function() {
```

Comment:
Add a comment or so to explain what is being tested in this nested `describe` block.